jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java

```diff
@@ -150,59 +155,80 @@ static Resource combine(Resource... resources)
     Resource newResource(URI uri);
 
     /**
-     * <p>Construct a system resource from a string.</p>
-     *
-     * <p>
-     * The resource is first attempted to be accessed via the {@link Thread#getContextClassLoader()}
-     * before being treated as a normal resource.
-     * </p>
+     * <p>Construct a Resource from a string reference into classloaders.</p>
      *
      * @param resource Resource as string representation
      * @return The new Resource, or null if string points to a location that does not exist
      * @throws IllegalArgumentException if string is blank
+     * @see #newClassLoaderResource(String, boolean)
+     * @deprecated use {@link #newClassLoaderResource(String)} or {@link #newClassLoaderResource(String, boolean)} instead, will be removed in Jetty 12.1.0
      */
+    @Deprecated(since = "12.0.2", forRemoval = true)
     default Resource newSystemResource(String resource)
+    {
+        return newClassLoaderResource(resource);
+    }
+
+    /**
+     * <p>Construct a Resource from a search of ClassLoaders.</p>
+     *
+     * <p>
+     *     Search order is:
+     * </p>
+     * <ol>
+     *   <li>{@link ClassLoader#getResource(String) java.lang.Thread.currentThread().getContextClassLoader().getResource(String)}</li>
+     *   <li>{@link ClassLoader#getResource(String) ResourceFactory.class.getClassLoader().getResource(String)}</li>
+     *   <li>(optional) {@link ClassLoader#getSystemResource(String) java.lang.ClassLoader.getSystemResource(String)}</li>
+     * </ol>
+     *
+     * <p>
+     *     See {@link ClassLoader#getResource(String)} for rules on resource name parameter.
+     * </p>
+     * 
+     * <p>
+     *     If a provided resource name starts with a {@code /} (example: {@code /org/example/ClassName.class})
+     *     then the non-slash version is also tried against the same ClassLoader (example: {@code org/example/ClassName.class}).
+     * </p>
+     *
+     * @param resource the resource name to find in a classloader
+     * @param searchSystemClassLoader true to search {@link ClassLoader#getSystemResource(String)}, false to skip
+     * @return The new Resource, or null if string points to a location that does not exist
+     * @throws IllegalArgumentException if resource name or resulting URL from ClassLoader is invalid.
+     */
+    default Resource newClassLoaderResource(String resource, boolean searchSystemClassLoader)
     {
         if (StringUtil.isBlank(resource))
             throw new IllegalArgumentException("Resource String is invalid: " + resource);
 
         URL url = null;
-        // Try to format as a URL?
-        ClassLoader loader = Thread.currentThread().getContextClassLoader();
-        if (loader != null)
+
+        List<Function<String, URL>> loaders = new ArrayList();
```

Comment:
I'm not sure Greg will like this allocation.

Reply:
It was a variation of his suggestion https://github.com/eclipse/jetty.project/pull/10533#discussion_r1331039547

Reply:
for the record, I don't like it, but does handle the iteration well. trust the JIT

Reply:
:+1: 
With a bit of luck, this `ArrayList` will be stack-allocated, hence garbage-free.